crux-agent/app/core/engine/self_evolve.py

```
@@ -294,6 +327,21 @@ async def _solve_internal(self, problem: Problem, start_iteration: int = 1, evol
             evolution_history.append(iteration_data)
 
+            # Save partial results to Redis if applicable
+            if self.job_id and self.redis_client:
+                partial_result = {
+                    "iterations": len(evolution_history),
+                    "latest_iteration": iteration_data,
+                    "evolution_history": evolution_history,
+                    # Include cumulative token metrics so the GUI can surface them mid-run
+                    "total_tokens": total_tokens,
+                    # Reasoning tokens are stored per-agent in metadata; aggregate if present
+                    "reasoning_tokens": total_reasoning_tokens if 'total_reasoning_tokens' in locals() else 0,
```

Ruff (0.12.2)

Undefined name `total_reasoning_tokens`

(F821)